optuna/samplers/_tpe/parzen_estimator.py

```diff
@@ -18,114 +33,419 @@ class _ParzenEstimatorParameters(
-                sigma[-1] = low_sorted_mus_high[-2] - low_sorted_mus_high[-3]
+            shape = (n_observations, len(choices))
+            assert prior_weight is not None
+            value = prior_weight / n_observations
+        weights = np.full(shape, fill_value=value)
+        weights[np.arange(n_observations), observations] += 1
+        weights /= weights.sum(axis=1, keepdims=True)
+        return weights
+
+    def _calculate_numerical_params(
+        self, observations: np.ndarray, param_name: str
+    ) -> Tuple[np.ndarray, np.ndarray]:
+
+        n_observations = self._n_observations
+        consider_prior = self._parameters.consider_prior
+        consider_magic_clip = self._parameters.consider_magic_clip
+        multivariate = self._parameters.multivariate
+        sigmas0 = self._sigmas0
+        low = self._low[param_name]
+        high = self._high[param_name]
+        assert low is not None
+        assert high is not None
+        assert len(observations) == self._n_observations
+
+        if n_observations == 0:
+            consider_prior = True
 
-        # We decide the weights.
-        unsorted_weights = weights_func(mus.size)
         if consider_prior:
-            sorted_weights = numpy.zeros_like(sorted_mus)
-            sorted_weights[:prior_pos] = unsorted_weights[order[:prior_pos]]
-            sorted_weights[prior_pos] = prior_weight
-            sorted_weights[prior_pos + 1 :] = unsorted_weights[order[prior_pos:]]
+
+            prior_mu = 0.5 * (low + high)
+            prior_sigma = 1.0 * (high - low)
+
+            mus = np.empty(n_observations + 1)
+            mus[:n_observations] = observations
+            mus[n_observations] = prior_mu
+
+            if not multivariate:
+                assert sigmas0 is None
+                pairs_of_observation_and_idx = np.asarray(
+                    [(low, -1)]
+                    + [(x, i) for i, x in enumerate(observations)]
+                    + [(high, n_observations + 1)]
+                    + [(prior_mu, n_observations)]
+                )
+                pairs_of_observation_and_idx = pairs_of_observation_and_idx[
+                    np.argsort(pairs_of_observation_and_idx[:, 0])
+                ]
+
+                pairs_of_sigma_and_idx = np.empty((n_observations + 1, 2))
+                pairs_of_sigma_and_idx[:, 0] = np.maximum(
+                    pairs_of_observation_and_idx[1:-1, 0] - pairs_of_observation_and_idx[0:-2, 0],
+                    pairs_of_observation_and_idx[2:, 0] - pairs_of_observation_and_idx[1:-1, 0],
+                )
+                pairs_of_sigma_and_idx[:, 1] = pairs_of_observation_and_idx[1:-1, 1]
+
+                sigmas = np.empty(n_observations + 1)
+                sigmas[:n_observations] = pairs_of_sigma_and_idx[
+                    np.argsort(pairs_of_sigma_and_idx[:, 1])
+                ][:-1, 0]
+                sigmas[n_observations] = prior_sigma
```

Comment:
Can we simplify this logic/datastructure? Some like https://github.com/optuna/optuna/commit/e2eeac76f3d3e9799ac0f432118171691b481969